sql/lease.go

```diff
@@ -0,0 +1,546 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/hlc"
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/cockroachdb/cockroach/util/retry"
+)
+
+// TODO(pmattis): Periodically renew leases for tables that were used recently and
+// for which the lease will expire soon.
+
+const (
+	leaseDuration    = int64(5 * time.Minute)
+	minLeaseDuration = int64(time.Minute)
+)
+
+const (
+	leaseDescIDRowIndex = iota
+	leaseVersionRowIndex
+	leaseExpirationRowIndex
+	leaseNodeIDRowIndex
+)
+
+var (
+	leaseValTypes          []parser.Datum
+	leaseColIDToRowIndex   = make(map[ColumnID]int, 4)
+	errLeaseVersionChanged = errors.New("lease version changed")
+)
+
+func init() {
+	columns := [...]string{
+		leaseDescIDRowIndex:     "descID",
+		leaseVersionRowIndex:    "version",
+		leaseExpirationRowIndex: "expiration",
+		leaseNodeIDRowIndex:     "nodeID",
+	}
+	for i, name := range columns {
+		col, err := LeaseTable.FindColumnByName(name)
+		if err != nil {
+			log.Fatal(err)
+		}
+		if col.ID == 0 {
+			// This shouldn't happen, just being paranoid.
+			log.Fatalf("column %s invalid", col.Name)
+		}
+		leaseColIDToRowIndex[col.ID] = i
+
+		// Verify the column is at the expected location in the primary index.
```

Comment:
Actually, this isn't necessary any more. It was necessary when I was decoding the lease rows, but I'm no longer doing so. I'll remove.